presto-main/src/main/java/com/facebook/presto/sql/gen/AndCodeGenerator.java

```diff
@@ -36,68 +38,74 @@ public BytecodeNode generateExpression(BytecodeGeneratorContext generator, Type
     {
         Preconditions.checkArgument(arguments.size() == 2);
 
-        Variable wasNull = generator.wasNull();
+        // We flatten the AND here.
+        Stack<RowExpression> stack = new Stack<RowExpression>();
+        stack.push(arguments.get(1));
+        stack.push(arguments.get(0));
+
+        List<RowExpression> flattenedArgs = new ArrayList<RowExpression>();
+        do {
+            RowExpression operand = stack.pop();
+            if (operand instanceof SpecialFormExpression &&
+                    ((SpecialFormExpression) operand).getForm() == SpecialFormExpression.Form.AND) {
+                stack.push(((SpecialFormExpression) operand).getArguments().get(1));
+                stack.push(((SpecialFormExpression) operand).getArguments().get(0));
+            }
+            else {
+                flattenedArgs.add(operand);
+            }
+        } while (!stack.empty());
+
         BytecodeBlock block = new BytecodeBlock()
                 .comment("AND")
                 .setDescription("AND");
 
-        BytecodeNode left = generator.generate(arguments.get(0), Optional.empty());
-        BytecodeNode right = generator.generate(arguments.get(1), Optional.empty());
-
-        block.append(left);
-
-        IfStatement ifLeftIsNull = new IfStatement("if left wasNull...")
-                .condition(wasNull);
-
-        LabelNode end = new LabelNode("end");
-        ifLeftIsNull.ifTrue()
-                .comment("clear the null flag, pop left value off stack, and push left null flag on the stack (true)")
-                .append(wasNull.set(constantFalse()))
-                .pop(arguments.get(0).getType().getJavaType()) // discard left value
-                .push(true);
+        LabelNode falseLabel = new LabelNode("false");
+        LabelNode endLabel = new LabelNode("end");
+        Variable wasNull = generator.wasNull();
 
-        LabelNode leftIsTrue = new LabelNode("leftIsTrue");
-        ifLeftIsNull.ifFalse()
-                .comment("if left is false, push false, and goto end")
-                .ifTrueGoto(leftIsTrue)
-                .push(false)
-                .gotoLabel(end)
-                .comment("left was true; push left null flag on the stack (false)")
-                .visitLabel(leftIsTrue)
+        // boolean hasNulls = false; do { eval arg; if (false) goto ret_false; if (wasNull) hasNulls = true; }
```

Comment:
It's too long. And also who can see the bytecode anyway. Also I gave comments on all important bytecodes.

Reply:
I don't understand the bytecode here. It seems to be different from what's described here? My read of the bytecode is
```
do {
  eval arg;
  if (wasNull) {
    hasNull = true;
    wasNull = false;
    pop eval result;
  }
  else {
    goto ret_false;
  }
}
```
which doesn't make sense to me...

Reply:
Also, can you end this loop early if `hasNull` is `true` already? Result will be null at that point right?

Reply:
No. That's the really bad part of SQL - for AND if any of them is false, you need to return false irrespective of other parts. But if none of them is false but some are null, you need to return go all the way and return null - similarly true for OR

Reply:
> I don't understand the bytecode here. It seems to be different from what's described here? My read of the bytecode is
> 
> ```
> do {
>   eval arg;
>   if (wasNull) {
>     hasNull = true;
>     wasNull = false;
>     pop eval result;
>   }
>   else {
>     goto ret_false;

Here else if (value_on_stack is false) { goto ret_false; }

otherwise, continue  
>   }
> }
> ```
> 
> which doesn't make sense to me...

Reply:
> No. That's the really bad part of SQL - for AND if any of them is false, you need to return false irrespective of other parts. But if none of them is false but some are null, you need to return go all the way and return null - similarly true for OR

😯

Reply:
Done